store/cachekv/store.go

```diff
@@ -279,73 +208,66 @@ func (store *Store) dirtyItems(start, end []byte) {
 		return
 	}
 
-	n := len(store.unsortedCache)
-	unsorted := make([]*kv.Pair, 0)
-	// If the unsortedCache is too big, its costs too much to determine
-	// whats in the subset we are concerned about.
-	// If you are interleaving iterator calls with writes, this can easily become an
-	// O(N^2) overhead.
-	// Even without that, too many range checks eventually becomes more expensive
-	// than just not having the cache.
-	if n < 1024 {
-		for key := range store.unsortedCache {
+	// dirty items needs to clear the unsorted items from the unsortedCache, then put them into the sortedCache.
+	// Subsequently an iterator is constructed using the sorted cache and parent store.
+	//
+	// However:
+	// if we have multiple iterators open, we have to avoid writes to the store.sortedCache MemDB.
+	// This is because a write to mem-db will definitely cause deadlocks in tm-db's memdb abstractions.
+	// (Though it could potentially be thread-safe per golang btree's docs)
+	// Thus we can only safely clear the unsorted cache to construct an iterator if we are:
+	// - only doing one iterator at a time
+	// - doing multiple iterators, over the same data, and you didn't write to the subset between
+	//   first iterator creation and current iterator creation
+	// - doing multiple iterators over disjoint pieces of data,
+	//   and have done no writes over any iterated component since first iterator creation
+	//
+	// To reword, in order to better accomodate multiple iterators,
+	// every time we construct an iterator when no others are active, we clear the entire unsorted cache.
+	// otherwise we don't do any clearing of the unsorted cache,
+	// and panic if we would be constructing an invalid iterator.
+	// TODO: Come back and vastly simplify the cacheKV store / iterator architecture,
```

Comment:
```suggestion
	// and panic if we would be constructing an invalid iterator.
	//
	// TODO: Come back and vastly simplify the cacheKV store / iterator architecture,
```

Let's create a GH issue and reference it here in the TODO. No naked TODO's :p